url10fb6bd43e_guwere_P2PSearch_tgz-pJ8-MainJ8/src/Controller.java

```diff
@@ -0,0 +1,577 @@
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+
+import java.io.IOException;
+import java.io.StringWriter;
+import java.net.DatagramPacket;
+import java.net.InetSocketAddress;
+import java.net.SocketException;
+import java.util.*;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+
+public final class Controller implements Runnable {
+
+    private int nodeID;
+    private BlockingQueue receiveQueue;// queues allow for complete decoupling of receiver<->controller<->sender
+    private BlockingQueue sendQueue;
+    private InetSocketAddress bootstrapNode;
+    private Map<Integer,Contact> routingTable;
+    private Indexer indexer;
+    //timers---
+    private Timer rTableTimer;
+    private Waiter searchResponseWaiter;
+    private Map<Integer,Timer> pingTimers;
+    private Map<String,Timer> indexTimers;
+    //----
+    private Vector<SearchResponse> resultResponses;//for each search invocation by the user the search responses are buffered before sent back
+    public Controller(BlockingQueue rQueue, BlockingQueue sQueue){
+         receiveQueue = rQueue;
+         sendQueue = sQueue;
+         rTableTimer = new Timer();
+        //start the routing table cleanup procedure after refreshTime on every refreshTime interval to weed out stale contacts
+         rTableTimer.scheduleAtFixedRate(new RefreshRoutingTable(),NetworkProperties.refreshTime*1000,NetworkProperties.refreshTime*1000);
+         routingTable = new ConcurrentHashMap<Integer, Contact>();//access from multiple threads
+         pingTimers = new ConcurrentHashMap<Integer, Timer>();
+        indexTimers = new ConcurrentHashMap<String, Timer>();
+
+    }
+
+    @Override
+    public void run() {
+        while(true){
+            try {
+            	DatagramPacket receivedPacket =(DatagramPacket)receiveQueue.take();
+            	//the queue blocks if empty
+                handleReceivedPacket(receivedPacket);
+            } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)